test/integration/garbage_collector_test.go

```diff
@@ -269,66 +289,90 @@ func TestCreateWithNonExisitentOwner(t *testing.T) {
 	}
 }
 
-func createRemoveRCsPods(t *testing.T, clientSet clientset.Interface, id int, wg *sync.WaitGroup, rcUIDs chan types.UID) {
+func setupRCsPods(t *testing.T, gc *garbagecollector.GarbageCollector, clientSet clientset.Interface, nameSuffix string, initialFinalizers []string, options *api.DeleteOptions, wg *sync.WaitGroup, rcUIDs chan types.UID) {
 	defer wg.Done()
 	rcClient := clientSet.Core().ReplicationControllers(framework.TestNS)
 	podClient := clientSet.Core().Pods(framework.TestNS)
 	// create rc.
-	rcName := toBeDeletedRCName + strconv.Itoa(id)
-	toBeDeletedRC, err := rcClient.Create(newOwnerRC(rcName))
+	rcName := "test.rc." + nameSuffix
+	rc := newOwnerRC(rcName)
+	rc.ObjectMeta.Finalizers = initialFinalizers
+	rc, err := rcClient.Create(rc)
 	if err != nil {
 		t.Fatalf("Failed to create replication controller: %v", err)
 	}
-	rcUIDs <- toBeDeletedRC.ObjectMeta.UID
-	// create pods. These pods should be cascadingly deleted.
+	rcUIDs <- rc.ObjectMeta.UID
+	// create pods.
+	var podUIDs []types.UID
 	for j := 0; j < 3; j++ {
-		podName := garbageCollectedPodName + strconv.Itoa(id) + "-" + strconv.Itoa(j)
-		pod := newPod(podName, []v1.OwnerReference{{UID: toBeDeletedRC.ObjectMeta.UID, Name: rcName}})
+		podName := "test.pod." + nameSuffix + "-" + strconv.Itoa(j)
+		pod := newPod(podName, []v1.OwnerReference{{UID: rc.ObjectMeta.UID, Name: rc.ObjectMeta.Name}})
 		_, err = podClient.Create(pod)
 		if err != nil {
 			t.Fatalf("Failed to create Pod: %v", err)
 		}
+		podUIDs = append(podUIDs, pod.ObjectMeta.UID)
+	}
+	orphan := (options != nil && options.OrphanDependents != nil && *options.OrphanDependents) || (options == nil && len(initialFinalizers) != 0 && initialFinalizers[0] == api.FinalizerOrphan)
+	// if we intend to orphan the pods, we need wait for the gc to observe the
+	// creation of the pods, otherwise if the deletion of RC is observed before
+	// the creation of the pods, the pods will not be orphaned.
```

Comment:
This sounds like it would be a breaking change in behavior. Why is it necessary? I would expect the actual condition you should wait for is the one that everyone currently uses, that the controller has made all of the pods and is at the correct generation...